compiler/formal_verification/src/typing.rs

```diff
@@ -310,201 +487,224 @@ pub fn type_infer(
+                            OptionalType::Well(NoirType::Tuple(types)) => {
+                                Some(TupleTypes::Well(types.clone()))
                             }
+                            OptionalType::PartialTuple(types) => {
+                                Some(TupleTypes::Partial(types.clone()))
+                            }
+                            _ => None,
+                        }
+                    }
 
-                            match &expr_left.ann.1 {
-                                // Fine shiftee
-                                OptionalType::Well(NoirType::Integer(_, _)) => {}
-                                // Integer literal, try type inferring to u32
-                                OptionalType::IntegerLiteral => {
-                                    expr_left = propagate_concrete_type(
-                                        expr_left,
-                                        default_literal_type.clone(),
-                                    )?;
-                                }
-                                // Not fine shiftee
-                                OptionalType::Well(expr_left_typ) => {
+                    if matches!(op, BinaryOp::Eq | BinaryOp::Neq)
+                        && let (Some(_), Some(_)) =
+                            (is_tupleish(&expr_left.ann.1), is_tupleish(&expr_right.ann.1))
+                    {
+                        let ExprF::Tuple { exprs: mut left_exprs } = *expr_left.expr.clone() else {
+                            unreachable!()
+                        };
+
+                        let ExprF::Tuple { exprs: mut right_exprs } = *expr_right.expr.clone()
+                        else {
+                            unreachable!()
+                        };
+
+                        if left_exprs.len() != right_exprs.len() {
+                            // TODO(totel): better error?
+                            return Err(TypeInferenceError::NoirTypeError(
+                                TypeCheckError::TupleMismatch {
+                                    location,
+                                    tuple_types: vec![],
+                                    actual_count: right_exprs.len(),
+                                },
+                            ));
+                        }
+
+                        let mut new_exprs_types = Vec::with_capacity(left_exprs.len());
+
+                        // NOTE: Unify all pair of elements
+                        for i in 0..left_exprs.len() {
+                            let (new_left_expr, new_right_expr, t) = unify(
+                                left_exprs[i].clone(),
+                                right_exprs[i].clone(),
+                                &default_literal_type,
+                            )?;
+
+                            (left_exprs[i], right_exprs[i]) = (new_left_expr, new_right_expr);
+                            new_exprs_types.push(t.clone());
+                        }
```

Comment:
```suggestion
                       let zipped = left_exprs.into_iter().zip(right_exprs.into_iter());

                        let (left_exprs, right_exprs): (Vec<_>, Vec<_>) = zipped
                            .map(|(left, right)| {
                                let (new_left, new_right, t) = unify(left, right, &default_literal_type)?;
                                new_exprs_types.push(t);
                                Ok((new_left, new_right))
                            })
                            .collect::<Result<Vec<_>, _>>()?
                            .into_iter()
                            .unzip();
```
You can change the snippet in the following way to avoid calling `clone()` on objects which you can consume.

Reply:
Resolved in an entirely different manner in the latest commit